File: generics/src/main/java/com/jthou/generics/AddTimerClassVisitor.java
//package com.jthou.generics;
//
//import org.objectweb.asm.ClassVisitor;
//import org.objectweb.asm.MethodVisitor;
//
//public class AddTimerClassVisitor extends ClassVisitor {
//
//    public AddTimerClassVisitor(int api, ClassVisitor classVisitor) {
//        super(api, classVisitor);
//    }
//
//    @Override
//    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
//
//        MethodVisitor methodVisitor = super.visitMethod(access, name, descriptor, signature, exceptions);
//
//        MethodVisitor newMethodVisitor = new MethodVisitor(api, methodVisitor) {
//
//        };
//
//        return newMethodVisitor;
//    }
//}

File: generics/src/main/java/com/jthou/generics/AddTimerClassVisitor.kt
package com.jthou.generics

import org.objectweb.asm.Opcodes
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor

class AddTimerClassVisitor : ClassVisitor {

    constructor(api: Int) : super(api)

    constructor(api: Int, classVisitor: ClassVisitor) : super(api, classVisitor)

    override fun visitEnd() {
        val fv = cv?.visitField(Opcodes.ACC_PUBLIC + Opcodes.ACC_STATIC, "timer", "J", null, null)
        fv?.visitEnd()
        cv?.visitEnd()
    }

    override fun visitMethod(
        access: Int,
        name: String?,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?
    ): MethodVisitor {
        val vm = super.visitMethod(access, name, descriptor, signature, exceptions)
        return object : MethodVisitor(api, vm) {}
    }

}